src/TransponderSquawk.hpp

```diff
@@ -0,0 +1,74 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+// Copyright The XCSoar Project
+
+#pragma once
+
+#include <cassert>
+#include <cstdint>
+
+/**
+ * This class stores a transponder squawk
+ */
+class TransponderSquawk {
+  int16_t value;
+
+public:
+  /**
+   * Uninitialized.
+   */
+  TransponderSquawk() = default;
+
+  /**
+   * Construct an empty instance.  Its IsDefined() method will return
+   * false.
+   */
+  static constexpr TransponderSquawk Null() noexcept {
+    return TransponderSquawk(-1);
+  }
+
+  constexpr explicit TransponderSquawk(int16_t value) {
+    SetCode(value);
+  }
+
+  constexpr void SetCode(int16_t code) {
+    value = !IsValidCode(code) ? -1 : code;
+  }
+
+  constexpr int16_t GetCode() const noexcept {
+    assert(IsDefined());
+
+    return value;
+  }
+
+  constexpr bool IsDefined() const noexcept {
+    return value != -1;
+  }
+
+  /**
+   * Set this object to "undefined".
+   */
+  constexpr void Clear() noexcept {
+    value = -1;
+  }
+
+private:
+  constexpr bool IsValidDigit(int8_t digit) const noexcept {
+    return digit >= 0 && digit <= 7;
+  }
+
+  constexpr bool IsValidCode(int16_t code) const noexcept {
+    if (code > 7777 || code < 0) {
+      return false;
+    }
+
+    // Check each digit for validity
+    for (int i = 0; i < 4; ++i) {
+      if (!IsValidDigit(code % 10)) {
+        return false;
+      }
+      code /= 10;
+    }
```

Comment:
This whole check looks awfully complicated and bloated, when you missed the fact that transponder codes are just octal values, and you could easily parse them as such with `strtoul(x, &endptr, 8)` and format with `%o` instead of pretending they are decimal.
With an octal representation, you can easily have:
```
static constexpr bool IsValidCode(uint_least16_t code) { return code <= 0777; }
```
Can it be any simpler? (and faster?)

Reply:
I recognized that transponder codes are octal, but I did not manage to make use of that realization. Your remarks were helpful, I hope the updated class looks better.